### PR TITLE
Add temperature-control cluster xml and generate zap

### PR DESCRIPTION
**Pull request by jadhavrohit924**
- Changes:

> Add temperature-control cluster xml and generate zap.

- Testing

> Build all-clusters-app, chip-tool

### CODE REVIEW

**Comment by andy31415** [comment]
@jadhavrohit924 lots of CI failures ... I think some file changes may be missing here.

**Comment by dhrishi** [comment]
> @jadhavrohit924 lots of CI failures ... I think some file changes may be missing here.

Yes, right. We will fix those

**Comment by jadhavrohit924** [comment]
@bzbarsky-apple Please review.

**Comment by jadhavrohit924** [comment]
@bzbarsky-apple I have implemented server code for temperature control cluster and added PR #26566. When I read attributes from chip-tool, chip-tool received correct command data. But it got message of `Don't know how to log atribute value`. Am I missing something here?

[chip-tool.txt](https://github.com/project-chip/connectedhomeip/files/11476310/chip-tool.txt)

**Comment by bzbarsky-apple** [comment]
@jadhavrohit924 How does this PR interact with the XML being added in https://github.com/project-chip/connectedhomeip/pull/26560 ?

Looking into your chip-tool issue

**Comment by bzbarsky-apple** [comment]
> Am I missing something here?

It's a chip-tool bug.  See https://github.com/project-chip/connectedhomeip/pull/26577

**Comment by jadhavrohit924** [comment]
@bzbarsky-apple 
> It's a chip-tool bug. See 26577.

Actually, I tried enabling client cluster in controller-clusters.zap. But it didn’t  work.

**Comment by bzbarsky-apple** [comment]
> Actually, I tried enabling client cluster in controller-clusters.zap. But it didn’t work.

I don't know exactly what you did, but if I enable the client cluster and then rerun codegen, DataModelLogger picks up these new attributes.  That's without https://github.com/project-chip/connectedhomeip/pull/26577.  With that it does not matter and it all just works.

File: src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:34]: Please take out the min and max bits for all of these readonly attributes: not only do they not match the spec, we're not using the values for anything.  And having them as positive hex values for negative things is fairly confusing.

File: src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:44]: Both of those args are optional, no?

File: src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:41]:
```suggestion
    <command source="client" code="0x0" name="SetTemperature" optional="false">
```

File: src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml
Comment (jadhavrohit924) [reply src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml]: Yes updated

File: src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml
Comment (jadhavrohit924) [reply src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml]: Removed

File: src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml
Comment (jadhavrohit924) [inline src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:44]: @bzbarsky-apple These shouldn’t be set to optional. Both of there args are feature dependent and cluster must have any one feature enabled. So if we set these args to optional, chip-tool will generate command like this `chip-tool temperaturecontrol set-temperature destination-id endpoint-id-ignored-for-group-commands`

Comment (jadhavrohit924) [reply src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:44]: We can definitely pass `TargetTemperature` and `TargetTemperatureLevel` to the command. But shouldn’t we have atleast one argument mandatory ( whichever feature we are enabling by default )?

Comment (bzbarsky-apple) [reply src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:44]: They are feature dependent, which means that any given cluster will have one or the other of them missing.  Which means the command struct must be able to represent "missing" so it does not send garbage on the wire.  Which is why they have to be optional.

The XML does not have a way to indicate "at least one of these is mandatory".  You have to define them both as optional and then consumers have to make sure they pass one or the other as needed.

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
@jadhavrohit924 Merge conflicts; needs a regen.

File: src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml
Comment (jadhavrohit924) [reply src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:44]: Okay. Thanks.

**Comment by woody-apple** [comment]
@jadhavrohit924 Would you mind taking a peek at the merge conflicts?

**Comment by bzbarsky-apple** [comment]
@woody-apple There are no merge conflicts; this is just blocked on non-Apple review.

**Review by dhrishi — APPROVED** [review]

File: src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:1]: Note that this is likely to need significant changes based on the outcome of https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/7040.  Please make sure those changes are either made before this lands or tracked in a cert blocker issue after this lands.

Comment (dhrishi) [reply src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:1]: @bzbarsky-apple Anything works. I can also also open a SDK issue and assign to @jadhavrohit924 to track this

Comment (dhrishi) [reply src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:1]: https://github.com/project-chip/connectedhomeip/issues/26646

**Review by jmartinez-silabs — APPROVED** [review]

File: src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml
Comment (marktrayer) [inline src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:28]: Please set domain to just: Appliances

Comment (jadhavrohit924) [reply src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml]: Done

**Comment by bzbarsky-apple** [comment]
This may need a rebase to tip to fix the ZAP CI issue, since it got updated while tip was broken... @jadhavrohit924

**Comment by dhrishi** [comment]
@jadhavrohit924  `Smoke test - Android / Smoke Run - Android` in CI is failing :
```
INFO    2023-05-19 02:12:02 INFO    Writing new data to: gen/src/controller/data_model/jni/FaultInjectionClient-InvokeSubscribeImpl.cpp
INFO    2023-05-19 02:12:02 CRITICAL expected and generated files do not match.
INFO    2023-05-19 02:12:02 CRITICAL    'jni/TemperatureControlClient-InvokeSubscribeImpl.cpp' was generated but not expected
INFO    2023-05-19 02:12:02 CRITICAL    'jni/TemperatureControlClient-ReadImpl.cpp' was generated but not expected
INFO    [60/1418] ACTION //src/controller/data_model:data_model_zapgen_zap_pregen(//build/toolchain/android:android_arm64)
INFO    [61/1418] ACTION //src/controller/java:tlv__kotlinc(//build/toolchain/android:android_arm64)
INFO    warning: flag is not supported by this version of the compiler: -Xlint:all
INFO    warning: flag is not supported by this version of the compiler: -Xlint:deprecation
INFO    ninja: build stopped: subcommand failed.
```
cc @bzbarsky-apple

**Comment by bzbarsky-apple** [comment]
@dhrishi @jadhavrohit924 Those two files need to be added to the outputs list in `src/controller/data_model/BUILD.gn`.  Search for `jni/` in there and note the list is in alphabetical order.

**Comment by jadhavrohit924** [comment]
@bzbarsky-apple Thanks for help :)

File: src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:23]: I missed this, but this does not match the spec....  I guess I'll fix it up in a separate PR, but @jadhavrohit924 please be more careful in the future.

Comment (bzbarsky-apple) [reply src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:23]: https://github.com/project-chip/connectedhomeip/pull/26691

Comment (dhrishi) [reply src/app/zap-templates/zcl/data-model/chip/temperature-control-cluster.xml:23]: @jadhavrohit924 Please take a note

**Pull request merged**
